xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -257,6 +278,29 @@ protected void handleLdsResponse(String versionInfo, List<Any> resources, String
     }
   }
 
+  private StructOrError<EnvoyServerProtoData.Listener> parseServerSideListener(Listener listener) {
+    String errorMessage = validateServerListener(listener);
+    if (errorMessage != null) {
+      return StructOrError.fromError(errorMessage);
+    }
+    try {
+      return StructOrError.fromStruct(EnvoyServerProtoData.Listener
+          .fromEnvoyProtoListener(listener));
+    } catch (InvalidProtocolBufferException e) {
+      return StructOrError.fromError(
+          "Failed to unpack Listener " + listener.getName() + ":" + e.getMessage());
+    }
+  }
+
+  private String validateServerListener(Listener listener) {
```

Comment:
You may just want to inline this helper method into `parseServerSideListener()`. Even later you add more validation logic to the Listener message, you probably want to _validate while parsing_ instead of doing this in two passes. That is, you will move `EnvoyServerProtoData.Listener.fromEnvoyProtoListener()`'s implementation into `parseServerSideListener()`.

Reply:
Yes, I realized that later and I will do that as part of this PR.